onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc

```diff
@@ -308,6 +378,130 @@ static bool MakeQDQNodeUnit(api::GraphRef& graph, const api::NodeRef& dq_node) {
   return true;
 }
 
+/// <summary>
+/// Fixes a Transpose QDQ node unit that is missing the DQ at its input.
+/// Inserts a Q -> DQ pair before the sequence Transpose -> Q by using the scale and zp info from the Q node.
+/// Before: prev_node (or graph input) ->                      Transpose -> Q
+/// After:  prev_node (or graph input) -> Q[new] -> DQ[new] -> Transpose -> Q
+/// </summary>
+/// <param name="transpose_node">Transpose node.</param>
+/// <param name="q_node">Q node.</param>
+/// <returns>True if Q -> DQ insertion was successful.</returns>
+static bool FixTransposeMissingDQ(api::GraphRef& graph, api::NodeRef& transpose_node, const api::NodeRef& q_node) {
+  assert(q_node.OpType() == "QuantizeLinear" && transpose_node.OpType() == "Transpose");
+  auto transpose_input_name = transpose_node.Inputs()[0];
+  auto transpose_output_name = transpose_node.Outputs()[0];
+  const auto q_node_inputs = q_node.Inputs();
+
+  if (transpose_output_name != q_node_inputs[0]) {
+    // Transpose and Q nodes are not connected. Should not happen.
+    return false;
+  }
+
+  auto transpose_output_consumers = graph.GetValueConsumers(transpose_output_name);
+  if (!transpose_output_consumers->comprehensive || transpose_output_consumers->nodes.size() != 1) {
+    // Q node should be the only consumer for the Transpose.
+    return false;
+  }
+
+  auto transpose_input_consumers = graph.GetValueConsumers(transpose_input_name);
+  if (transpose_input_consumers->nodes.size() != 1) {
+    // The transpose node should be the only consumer of its own input.
+    return false;
+  }
+
+  if (graph.GetConstant(transpose_input_name) != nullptr) {
+    // Don't handle a constant input to transpose node. This should have been constant folded.
```

Comment:
I'm not sure I'm fully understanding. Constant folding runs before we call this QDQ fix up function, so we shouldn't see (initializer -> Transpose -> Q). And if we do, this function does not handle it, so we exit anyway.

Reply:
I was thinking about the L1 constant folding optimizer. It's slightly confusing to have 'constant folding' meaning two different things. Maybe this could say `TryConstantFoldNode should have handled this` to clarify?